// angular-9-datatable/esm5/lib/paginator.component.js
import { __decorate, __metadata, __param } from "tslib";
import { Component, Input, Optional } from "@angular/core";
import { DataTable, PageEvent } from "./data-table.directive";
var Paginator = /** @class */ (function () {
    function Paginator(injectMfTable) {
        var _this = this;
        this.injectMfTable = injectMfTable;
        this.dataLength = 0;
        this.onPageChangeSubscriber = function (event) {
            _this.activePage = event.activePage;
            _this.rowsOnPage = event.rowsOnPage;
            _this.dataLength = event.dataLength;
            _this.lastPage = Math.ceil(_this.dataLength / _this.rowsOnPage);
        };
    }
    Paginator.prototype.ngOnChanges = function (changes) {
        this.mfTable = this.inputMfTable || this.injectMfTable;
        this.onPageChangeSubscriber(this.mfTable.getPage());
        this.mfTable.onPageChange.subscribe(this.onPageChangeSubscriber);
    };
    Paginator.prototype.setPage = function (pageNumber) {
        this.mfTable.setPage(pageNumber, this.rowsOnPage);
    };
    Paginator.prototype.setRowsOnPage = function (rowsOnPage) {
        this.mfTable.setPage(this.activePage, rowsOnPage);
    };
    Paginator.ctorParameters = function () { return [
        { type: DataTable, decorators: [{ type: Optional }] }
    ]; };
    __decorate([
        Input("mfTable"),
        __metadata("design:type", DataTable)
    ], Paginator.prototype, "inputMfTable", void 0);
    Paginator = __decorate([
        Component({
            selector: "mfPaginator",
            template: "<ng-content></ng-content>"
        }),
        __param(0, Optional()),
        __metadata("design:paramtypes", [DataTable])
    ], Paginator);
    return Paginator;
}());
export { Paginator };
